sinks/cache/in_memory.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+type entry struct {
+	timestamp time.Time
+	data      interface{}
+}
+
+// TODO: Consider using cadvisor's in memory storage instead.
+type timeStore struct {
+	buffer *list.List
+	rwLock sync.RWMutex
+}
+
+func (ts *timeStore) Put(timestamp time.Time, data interface{}) error {
+	if data == nil {
+		return fmt.Errorf("cannot store nil data")
+	}
+	ts.rwLock.Lock()
+	defer ts.rwLock.Unlock()
+	if ts.buffer.Len() == 0 {
+		glog.V(5).Infof("put pushfront: %v, %v", timestamp, data)
+		ts.buffer.PushFront(entry{timestamp: timestamp, data: data})
+		return nil
+	}
+	for elem := ts.buffer.Front(); elem != nil; elem = elem.Next() {
+		if timestamp.After(elem.Value.(entry).timestamp) {
+			glog.V(5).Infof("put insert before: %v, %v, %v", elem, timestamp, data)
+			ts.buffer.InsertBefore(entry{timestamp: timestamp, data: data}, elem)
+			return nil
+		}
+	}
+	glog.V(5).Infof("put pushback: %v, %v", timestamp, data)
+	ts.buffer.PushBack((entry{timestamp: timestamp, data: data}))
+	return nil
+}
+
+func (ts *timeStore) GetAll() []interface{} {
+	ts.rwLock.RLock()
+	defer ts.rwLock.RUnlock()
+	result := []interface{}{}
+	for elem := ts.buffer.Back(); elem != nil; elem = elem.Prev() {
+		result = append(result, elem.Value.(entry).data)
+	}
+
+	return result
+}
+
+func (ts *timeStore) Get(start, end time.Time) []interface{} {
+	ts.rwLock.RLock()
+	defer ts.rwLock.RUnlock()
+	if ts.buffer.Len() == 0 {
+		return nil
+	}
+	nullTime := time.Time{}
+	if start == nullTime {
+		start = time.Unix(0, 0)
+	}
+	result := []interface{}{}
+	for elem := ts.buffer.Back(); elem != nil; elem = elem.Prev() {
+		entry := elem.Value.(entry)
+		if entry.timestamp.Before(start) || (end != nullTime && entry.timestamp.After(end)) {
```

Comment:
I think there is an issue here with start time.

```
Take: [3, 2, 1]
Get(2, 5)

I would expect it to return [2, 3], but it actually returns []

We check:
- is 1 before 2? Yes, break.
```
Issue with [4, 3, 2, 1,] I want 2 - 3?